File: src/051/p051.py
import itertools
from src.utils.py.prime import is_prime


def generate_family(num: str, mask: tuple[int, int, int]) -> list[str]:
    family: list[str] = []
    digits = range(1, 10) if 0 in mask else range(10)
    for digit in digits:
        family.append(
            "".join(str(digit) if i in mask else n for i, n in enumerate(num))
        )
    return family


def generate_prime_family(num: str, primes: list[int]) -> list[str]:
    family: list[str] = []
    for indices in itertools.combinations(range(len(num) - 1), 3):
        if num[indices[0]] == num[indices[1]] == num[indices[2]]:
            candidates = generate_family(num, indices)
            for candidate in candidates:
                if is_prime(int(candidate), primes):
                    family.append(candidate)
    return family


def solve() -> int:
    sieve: dict[int, list[int]] = {}
    primes: list[int] = []

    # pylint: disable=duplicate-code
    for num in itertools.count(start=2):
        if num in sieve:
            # num is composite and sieve[num] contains its prime factors
            factors = sieve[num]
            del sieve[num]
        else:
            # num is prime
            factors = [num]
            primes.append(num)
            family = generate_prime_family(str(num), primes)
            if len(family) == 8:
                return num

        # Mark next multiple of each factor
        for prime in factors:
            sieve.setdefault(num + prime, []).append(prime)
    # pylint: enable=duplicate-code
    raise RuntimeError("No solution found")

File: src/P051/p051.go
package problem051

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/strutils"
)

var digits = []string{"1", "2", "3", "4", "5", "6", "7", "8", "9"}

func generatePrimeFamily(n int, primes []int) []int {
	str := strconv.Itoa(n)
	for i := 0; i < len(str)-1; i++ {
		for j := i + 1; j < len(str)-1; j++ {
			for k := j + 1; k < len(str)-1; k++ {
				if str[i] == str[j] && str[j] == str[k] {
					family := []int{}
					for _, d := range digits {
						candidate := str[:i] + d + str[i+1:j] + d + str[j+1:k] + d + str[k+1:]
						if p := strutils.Atoi(candidate); primeutils.IsPrime(p, primes) {
							family = append(family, p)
						}
					}
					if len(family) == 8 {
						return family
					}
				}
			}
		}
	}
	return nil
}

func Solve() string {
	sieve := make(map[int][]int)
	primes := []int{}

	for n := 2; ; n++ {
		var factors []int
		if _, exists := sieve[n]; exists {
			factors = sieve[n]
			delete(sieve, n)
		} else {
			factors = []int{n}
			primes = append(primes, n)
			if family := generatePrimeFamily(n, primes); family != nil {
				return strconv.Itoa(n)
			}
		}

		for _, f := range factors {
			sieve[n+f] = append(sieve[n+f], f)
		}
	}
}
